core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,56 +125,76 @@ public class JsPeripheral internal constructor(
     private val gatt: BluetoothRemoteGATTServer
         get() = bluetoothDevice.gatt!! // fixme: !!
 
-    private var connectJob: Deferred<Unit>? = null
-
-    private fun connectAsync() = scope.async(start = LAZY) {
-        logger.info { message = "Connecting" }
-        _state.value = State.Connecting.Bluetooth
-
+    private val connectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
         try {
-            registerDisconnectedListener()
-
-            gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
-            _state.value = State.Connecting.Services
-
-            discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            disconnectJob.launchAsync()
             throw t
         }
-
-        logger.info { message = "Connected" }
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private val disconnectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
+        connectJob.join()
+        closeConnection()
     }
 
     public override suspend fun connect() {
-        val job = connectJob ?: connectAsync().also { connectJob = it }
-        job.await()
+        connectJob.launchAsync().await()
     }
 
     public override suspend fun disconnect() {
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        connectJob.cancel()
+        disconnectJob.launchAsync().await()
+    }
+
+    private suspend fun openConnection() {
+        // Guard if already connected do nothing
+        if (_state.value is State.Connected) return
+
+        logger.info { message = "Connecting" }
+        // Wait until fully disconnected before proceeding
+        _state.first { it is State.Disconnected }
```

Comment:
Would `disconnectJob.join()` accomplish this?

Reply:
Probably "yes", but I think it may not be a good idea.

I've deliberately kept the job-to-job interactions out of these `closeConnection` and `openConnection` functions. In my mind this is "logic layer" and is distinct from the invoking async job layer. The logic here is easier to follow if we rely on the `_state` only.

Introducing the invoking job objects compounds the method interdependency and makes it harder to reason about the code IMHO.